### ISSUE TITLE
Problems with chgheader

### ISSUE THREAD

**Issue by nowrap** [Atlantis-Software/milterjs#4]
Hello,
we are trying to build a smime milter with milterjs.

Adding new headers works without problems.
But changing already set headers doesn't work for us right now and throws warnings in our postfix mta. (within eom event)

Trying to delete:
postfix/cleanup[21406]: warning: milter inet:127.0.0.1:9568: null change header name
ctx.chgheader('Content-Type', 0, '');

Trying to change:
postfix/cleanup[21480]: warning: milter8_read_data: left-over data 114 bytes
ctx.chgheader('Content-Type', 0, 'multipart/signed; protocol="application/pkcs7-signature"; micalg=sha256; boundary="--_NmP-a56176f7da52d5ca-Part_1"');

Maybe you have a clue?

Regards
nowrap

**Comment by atiertant** [comment]
Hi,

what's your postfix version ?

did you tried sniffing with wireshark to see if milterjs looks correct with milter protocol ?
see http://www.sendmail.org/~ca/email/sm-X/design-2005-05-05/main/node5.html#SECTION00591000000000000000

and compare with a working milter...

for change, have you tried a shorter text ?

**Comment by nowrap** [comment]
Hi atiertant,
thanks for your reply.

Today we found the source of the issue.
Postfix version, milter protocol version or length of the value weren't the problems on our ubuntu 16 lts system.

We tried to compare with other milter sources: python, perl and java.
Finally a zimbra milter class pointed us in the right direction:
https://github.com/fciubotaru/z-pec/blob/master/ZimbraServer/src/java/com/zimbra/cs/milter/MilterHandler.java

The chgheader was slighty different:
1.) 4 bits for the unsigned int 32
2.) headername + terminator
3.) headervalue + terminator

So the offset in your code needed a small change.
From:
`var data = Buffer.alloc(6 + header.length + value.length);`
`data.writeUInt32BE(num, 0);`
`data.write(header + '\0' + value + '\0', 5);`
To:
`var data = Buffer.alloc(6 + header.length + value.length);`
`data.writeUInt32BE(num, 0);`
`data.write(header + '\0' + value + '\0', 4);`

There was a terminator between the index and the headername. So a check in the postfix milter source got triggered.

Regards
nowrap
